File: docs/sdk/payclient.go
// Package payclient 是「统一支付中心」的 Go 接入 SDK（零依赖，仅用标准库 net/http，Go 1.18+）
//
// 用法：
//
//	client := payclient.New("https://pay.example.com", appID, appSecret)
//	order, err := client.CreateOrder(map[string]any{
//	    "merchantOrderNo": "ORDER_001",
//	    "amount":          "128.50",
//	    "subject":         "会员年卡",
//	    "tradeType":       "NATIVE",
//	})
//
// 约定：
//   - 金额一律字符串（元，两位小数），禁止浮点运算
//   - 所有写接口带幂等键（merchantOrderNo / merchantRefundNo），重试安全
//   - code != 0 返回 *PayError（含 Code / TraceID）
package payclient

import (
	"bytes"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// 开放接口路径
const (
	PathPayCreate    = "/api/v1/open/pay/create"
	PathPayQuery     = "/api/v1/open/pay/query"
	PathPayClose     = "/api/v1/open/pay/close"
	PathPayChannels  = "/api/v1/open/pay/channels"
	// ChannelAuto 渠道自动路由：不传 channel 时由支付中心自动选择，新增渠道无需升级 SDK
	ChannelAuto = "auto"
	PathRefundCreate = "/api/v1/open/refund/create"
	PathRefundQuery  = "/api/v1/open/refund/query"
)

// Client 支付中心客户端
type Client struct {
	BaseURL   string
	AppID     string
	AppSecret string
	Timeout   time.Duration
	// 如需自定义代理 / 连接池，可替换该字段
	HTTPClient *http.Client
}

// New 创建客户端，默认超时 10 秒
func New(baseURL, appID, appSecret string) *Client {
	return &Client{
		BaseURL:   strings.TrimRight(baseURL, "/"),
		AppID:     appID,
		AppSecret: appSecret,
		Timeout:   10 * time.Second,
	}
}

// PayError 支付中心返回的业务错误（code != 0）
type PayError struct {
	Code    int
	Message string
	TraceID string
}

func (e *PayError) Error() string {
	return fmt.Sprintf("[%d] %s", e.Code, e.Message)
}

// ==================== 开放接口 ====================

// CreateOrder 下单（幂等：同一 merchantOrderNo 返回同一订单）；未指定 channel 时由服务端路由
func (c *Client) CreateOrder(params map[string]any) (map[string]any, error) {
	body := make(map[string]any, len(params)+1)
	for k, v := range params {
		body[k] = v
	}
	if _, ok := body["channel"]; !ok {
		body["channel"] = ChannelAuto
	}
	return c.post(PathPayCreate, body)
}

// ListChannels 查询当前应用可用渠道，新增渠道会自动出现，无需升级 SDK
func (c *Client) ListChannels() (map[string]any, error) {
	return c.post(PathPayChannels, map[string]any{"appId": c.AppID})
}

// QueryOrder 查单：{"payOrderNo": ...} 或 {"merchantOrderNo": ...}
func (c *Client) QueryOrder(params map[string]any) (map[string]any, error) {
	return c.post(PathPayQuery, params)
}

// CloseOrder 关闭未支付订单
func (c *Client) CloseOrder(payOrderNo string) (map[string]any, error) {
	return c.post(PathPayClose, map[string]any{"payOrderNo": payOrderNo})
}

// Refund 退款（幂等键 merchantRefundNo）
func (c *Client) Refund(params map[string]any) (map[string]any, error) {
	return c.post(PathRefundCreate, params)
}

// QueryRefund 查退款：{"refundNo": ...} 或 {"merchantRefundNo": ...}
func (c *Client) QueryRefund(params map[string]any) (map[string]any, error) {
	return c.post(PathRefundQuery, params)
}

// ==================== 通知验签 ====================

// VerifyNotify 校验支付中心推送的通知签名（务必在业务落库前校验）
// headers 为原始请求头（key 大小写不敏感），rawBody 为未经解析的原始 body 字符串
func (c *Client) VerifyNotify(headers map[string]string, rawBody string) bool {
	ts := header(headers, "x-pay-timestamp")
	nonce := header(headers, "x-pay-nonce")
	sign := header(headers, "x-pay-sign")
	if ts == "" || nonce == "" || sign == "" {
		return false
	}
	appID := header(headers, "x-pay-app-id")
	if appID == "" {
		appID = c.AppID
	}

	sum := sha256.Sum256([]byte(rawBody))
	content := strings.Join([]string{appID, ts, nonce, hex.EncodeToString(sum[:])}, "\n")
	expected := c.hmacSha256Hex(content)
	return subtle.ConstantTimeCompare([]byte(expected), []byte(strings.ToLower(sign))) == 1
}

// ==================== 签名 ====================

// BuildSignHeaders 构造签名请求头
// 待签串 = appId \n timestamp \n nonce \n METHOD \n path \n sha256(body)
func (c *Client) BuildSignHeaders(method, path, body string) map[string]string {
	timestamp := strconv.FormatInt(time.Now().Unix(), 10)
	nonce := randomHex(12)
	sum := sha256.Sum256([]byte(body))
	content := strings.Join([]string{
		c.AppID, timestamp, nonce, strings.ToUpper(method), path, hex.EncodeToString(sum[:]),
	}, "\n")

	return map[string]string{
		"X-App-Id":    c.AppID,
		"X-Timestamp": timestamp,
		"X-Nonce":     nonce,
		"X-Sign":      c.hmacSha256Hex(content),
	}
}

// ==================== 内部实现 ====================

func (c *Client) post(path string, payload map[string]any) (map[string]any, error) {
	if payload == nil {
		payload = map[string]any{}
	}
	if _, ok := payload["appId"]; !ok {
		payload["appId"] = c.AppID
	}
	// 参与签名与发送的必须是同一份字节
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, &PayError{Code: -1, Message: "请求体序列化失败：" + err.Error()}
	}

	req, err := http.NewRequest(http.MethodPost, c.BaseURL+path, bytes.NewReader(body))
	if err != nil {
		return nil, &PayError{Code: -1, Message: "构造请求失败：" + err.Error()}
	}
	req.Header.Set("Content-Type", "application/json")
	for k, v := range c.BuildSignHeaders(http.MethodPost, path, string(body)) {
		req.Header.Set(k, v)
	}

	resp, err := c.http().Do(req)
	if err != nil {
		return nil, &PayError{Code: -1, Message: "请求支付中心失败：" + err.Error()}
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &PayError{Code: -1, Message: "读取响应失败：" + err.Error()}
	}

	var out struct {
		Code    int            `json:"code"`
		Message string         `json:"message"`
		Data    map[string]any `json:"data"`
		TraceID string         `json:"traceId"`
	}
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, &PayError{Code: -1, Message: "响应解析失败：" + string(raw)}
	}
	if out.Code != 0 {
		return nil, &PayError{Code: out.Code, Message: out.Message, TraceID: out.TraceID}
	}
	if out.Data == nil {
		out.Data = map[string]any{}
	}
	return out.Data, nil
}

func (c *Client) http() *http.Client {
	if c.HTTPClient != nil {
		return c.HTTPClient
	}
	timeout := c.Timeout
	if timeout <= 0 {
		timeout = 10 * time.Second
	}
	return &http.Client{Timeout: timeout}
}

func (c *Client) hmacSha256Hex(content string) string {
	mac := hmac.New(sha256.New, []byte(c.AppSecret))
	mac.Write([]byte(content))
	return hex.EncodeToString(mac.Sum(nil))
}

func randomHex(n int) string {
	buf := make([]byte, n)
	if _, err := rand.Read(buf); err != nil {
		return strconv.FormatInt(time.Now().UnixNano(), 16)
	}
	return hex.EncodeToString(buf)
}

func header(headers map[string]string, lowerName string) string {
	if v, ok := headers[lowerName]; ok {
		return v
	}
	for k, v := range headers {
		if strings.EqualFold(k, lowerName) {
			return v
		}
	}
	return ""
}

File: docs/sdk/pay_client.py
"""统一支付中心 · Python 接入 SDK（零依赖，仅用标准库 urllib，Python 3.8+）

用法：
    from pay_client import PayClient, PayError

    client = PayClient("https://pay.example.com", app_id, app_secret)
    order = client.create_order({
        "merchantOrderNo": "ORDER_001",
        "amount": "128.50",
        "subject": "会员年卡",
        "tradeType": "NATIVE",
    })

约定：
    - 金额一律字符串（元，两位小数），禁止浮点运算
    - 所有写接口带幂等键（merchantOrderNo / merchantRefundNo），重试安全
    - code != 0 抛出 PayError（含 code / trace_id）
"""

import hashlib
import hmac
import json
import secrets
import time
import urllib.error
import urllib.request
from typing import Any, Dict, Mapping, Optional


class PayError(Exception):
    """支付中心返回的业务异常（code != 0）"""

    def __init__(self, code: int, message: str, trace_id: Optional[str] = None):
        super().__init__(f"[{code}] {message}")
        self.code = code
        self.message = message
        self.trace_id = trace_id


class PayClient:
    PATH_PAY_CREATE = "/api/v1/open/pay/create"
    PATH_PAY_QUERY = "/api/v1/open/pay/query"
    PATH_PAY_CLOSE = "/api/v1/open/pay/close"
    PATH_REFUND_CREATE = "/api/v1/open/refund/create"
    PATH_REFUND_QUERY = "/api/v1/open/refund/query"

    def __init__(self, base_url: str, app_id: str, app_secret: str, timeout: float = 10.0):
        self.base_url = base_url.rstrip("/")
        self.app_id = app_id
        self.app_secret = app_secret
        self.timeout = timeout

    # ==================== 开放接口 ====================

    def create_order(self, params: Mapping[str, Any]) -> Dict[str, Any]:
        """下单（幂等：同一 merchantOrderNo 返回同一订单）"""
        return self._post(self.PATH_PAY_CREATE, params)

    def query_order(self, params: Mapping[str, Any]) -> Dict[str, Any]:
        """查单：{"payOrderNo": ...} 或 {"merchantOrderNo": ...}"""
        return self._post(self.PATH_PAY_QUERY, params)

    def close_order(self, pay_order_no: str) -> Dict[str, Any]:
        """关闭未支付订单"""
        return self._post(self.PATH_PAY_CLOSE, {"payOrderNo": pay_order_no})

    def refund(self, params: Mapping[str, Any]) -> Dict[str, Any]:
        """退款（幂等键 merchantRefundNo）"""
        return self._post(self.PATH_REFUND_CREATE, params)

    def query_refund(self, params: Mapping[str, Any]) -> Dict[str, Any]:
        """查退款：{"refundNo": ...} 或 {"merchantRefundNo": ...}"""
        return self._post(self.PATH_REFUND_QUERY, params)

    # ==================== 通知验签 ====================

    def verify_notify(self, headers: Mapping[str, str], raw_body: str) -> bool:
        """校验支付中心推送的通知签名（务必在业务落库前校验）

        :param headers: 原始请求头（key 大小写不敏感）
        :param raw_body: 未经任何解析的原始 body 字符串
        """
        ts = self._header(headers, "x-pay-timestamp")
        nonce = self._header(headers, "x-pay-nonce")
        sign = self._header(headers, "x-pay-sign")
        app_id = self._header(headers, "x-pay-app-id") or self.app_id
        if not ts or not nonce or not sign:
            return False

        content = "\n".join([app_id, ts, nonce, self._sha256_hex(raw_body or "")])
        expected = self._hmac_sha256_hex(self.app_secret, content)
        return hmac.compare_digest(expected, sign.lower())

    # ==================== 签名 ====================

    def build_sign_headers(self, method: str, path: str, body: str) -> Dict[str, str]:
        """待签串 = appId \\n timestamp \\n nonce \\n METHOD \\n path \\n sha256(body)"""
        timestamp = str(int(time.time()))
        nonce = secrets.token_hex(12)
        content = "\n".join(
            [self.app_id, timestamp, nonce, method.upper(), path, self._sha256_hex(body)]
        )
        return {
            "X-App-Id": self.app_id,
            "X-Timestamp": timestamp,
            "X-Nonce": nonce,
            "X-Sign": self._hmac_sha256_hex(self.app_secret, content),
        }

    # ==================== 内部实现 ====================

    def _post(self, path: str, payload: Optional[Mapping[str, Any]]) -> Dict[str, Any]:
        body_map: Dict[str, Any] = dict(payload or {})
        body_map.setdefault("appId", self.app_id)
        # 参与签名与发送的必须是同一个字符串
        body = json.dumps(body_map, ensure_ascii=False, separators=(",", ":"))

        req = urllib.request.Request(
            self.base_url + path,
            data=body.encode("utf-8"),
            method="POST",
            headers={"Content-Type": "application/json", **self.build_sign_headers("POST", path, body)},
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                raw = resp.read().decode("utf-8")
        except urllib.error.HTTPError as e:
            raw = e.read().decode("utf-8", "ignore")
        except Exception as e:  # 网络层异常
            raise PayError(-1, f"请求支付中心失败：{e}")

        try:
            resp_json = json.loads(raw)
        except Exception:
            raise PayError(-1, f"响应解析失败：{raw}")

        code = int(resp_json.get("code", -1))
        if code != 0:
            raise PayError(code, resp_json.get("message", "支付中心返回失败"), resp_json.get("traceId"))
        data = resp_json.get("data")
        return data if isinstance(data, dict) else {}

    @staticmethod
    def _header(headers: Mapping[str, str], lower_name: str) -> str:
        if headers is None:
            return ""
        for k, v in headers.items():
            if isinstance(k, str) and k.lower() == lower_name:
                return str(v or "")
        return ""

    @staticmethod
    def _sha256_hex(s: str) -> str:
        return hashlib.sha256(s.encode("utf-8")).hexdigest()

    @staticmethod
    def _hmac_sha256_hex(key: str, msg: str) -> str:
        return hmac.new(key.encode("utf-8"), msg.encode("utf-8"), hashlib.sha256).hexdigest()
